### DISCUSSION TITLE
Initialize provider in initState of a Widget

### DISCUSSION THREAD

**Discussion by lisbejuin** [rrousselGit/riverpod#2817] [Q&A]
Hello,
I have a problem with my provider initialization.
I have a form in a widget with an object who have a value chosen by user. The choice of this value is possible by a list of diverses other values.
=> WidgetA = form with 'chosenValue' and button 'Edit'
=> when clic on Edit button a have a list with divers values and the 'choosenValue' is checked, this list is associated with a state provider 'StateProvider1'
=> I want to initialize this state with 'choosenValue' in the 'initState' of my WidgetA and not with 'family'
When I do this in 'initState' I have this error 'Tried to modify a provider while the widget tree was building.'
There is a solution of my problem please ?
Thanks for your reply

**Comment by AhmedLSayed9** [comment]
Can you share a sample code? we can't help without any further information :)

**Reply by lisbejuin** [reply]
I have an object PatternInput, it's object of the form :
```
@freezed
class PatternInput with _$PatternInput {
  const factory PatternInput({
    required String id,
    required String brand,
  }) = _PatternInput;
}
```
there is my form:
```
class PatternDetailsPage extends ItemDetailPage<PatternInput> {
  PatternDetailsPage(String title, PatternInput item, Key? key})
      : super(title, item, key: key);

  @override
  ConsumerState<ConsumerStatefulWidget> createState() => _PatternDetailsPageState();
}

class _PatternDetailsPageState extends ItemDetailPageState<PatternInput> {
  @override
  void initState() {
    super.initState();
    ref.read(userChoiceProvider.notifier).addUserChoices(itemInput.categories);
  }

@override
  Widget build(BuildContext context) {
    return Scaffold(
        body: Row (children : [
              Text($item.brand),
              ElevatedButton(
                   onPressed: () { 
                           context.pushRoute(
                                   ListOfChoicesRoute()
                            );
                   },
                   child: Text('Add Brand'),
              ),
            ],
         ),
      ),
  }
```

There is the ListOfChoicePage
```
class ListOfChoicesPage extends ConsumerStatefulWidget {

  const ListOfChoicesPage({Key? key}) : super(key: key);

  @override
  ConsumerState<ConsumerStatefulWidget> createState() => _ListOfChoicesPageState();
}

class _ListOfChoicesPageState extends ConsumerState<ListOfChoicesPage> {

  @override
  Widget build(BuildContext context) {
          final userBrandChoice = ref.watch(userChoiceProvider);
          final choices = ref.watch(choicesProvider);
          return Scaffold(
          body: AppListView(
               itemCount: choices.length,
               itemBuilder: (BuildContext context, int index) {
                     ListChoice item = choices.values.elementAt(index);
                     bool isChoosen = choices.contains(userBrandChoice);
                     return AppListTile.multipleChoice(
                            title: item.title,
                            checkboxValue: isChoosen,
                     );
               },
          ),
          );
  }
}
```

There is my userChoiceProvider:
```
@riverpod
class UserChoice extends _$UserChoice {
  @override
  String build() {
    return '';
  }

  void addUserChoices(String value) {
    state = value;
  }
}

```

**Reply by lisbejuin** [reply]
I try to simplify my code to maximum :)

**Reply by AhmedLSayed9** [reply]
You cannot modify providers while widgets are building. Instead, do that inside the `onPressed` callback before pushing your route.

**Reply by lisbejuin** [reply]
Yes I know and I already try this solution but it doesn’t work… I think is a thread problem because when I run we debug and a breakpoint it works but when I run normally the value is not set when the list  opens

**Reply by AhmedLSayed9** [reply]
Mutating providers inside `onPressed` is working fine.
Maybe there's something else related to your code.

**Reply by lisbejuin** [reply]
Ok I understood the problem, when I push the new route the provider is automatically disposed before the new view is loaded so when the new view access to the provider, this one is reset... I solved my problem with @Riverpod(keepAlive:true) annotation on provider and I dispose it manually when I dismiss the list.

**Reply by lisbejuin** [reply]
thanks for your help :)
